File: thea-tizen/src/services/ai/OpenRouterProvider.ts
/**
 * OpenRouter Provider Implementation
 * OpenRouter provides unified access to multiple AI models
 */

import { BaseAIProvider, ProviderCapability, ProviderError } from './AIProvider';
import type { ChatMessage, ChatOptions, StreamChunk, AIModel } from '../../types/chat';
import { API_URLS, STREAM_CONFIG, APP_INFO } from '../../config/constants';

/**
 * OpenRouter provider - unified API for multiple AI models
 */
export class OpenRouterProvider extends BaseAIProvider {
  readonly id = 'openrouter';
  readonly name = 'OpenRouter';

  readonly capabilities = new Set<ProviderCapability>([
    'chat',
    'streaming',
    'vision',
    'functionCalling',
  ]);

  readonly supportedModels: AIModel[] = [
    // Anthropic models via OpenRouter
    {
      id: 'anthropic/claude-sonnet-4',
      name: 'Claude Sonnet 4',
      provider: 'openrouter',
      contextWindow: 200000,
      maxOutputTokens: 64000,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'reasoning', 'code'],
      inputCostPer1K: 0.003,
      outputCostPer1K: 0.015,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    {
      id: 'anthropic/claude-opus-4',
      name: 'Claude Opus 4',
      provider: 'openrouter',
      contextWindow: 200000,
      maxOutputTokens: 32000,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'reasoning', 'code'],
      inputCostPer1K: 0.015,
      outputCostPer1K: 0.075,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    {
      id: 'anthropic/claude-3.5-haiku',
      name: 'Claude 3.5 Haiku',
      provider: 'openrouter',
      contextWindow: 200000,
      maxOutputTokens: 8192,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'code'],
      inputCostPer1K: 0.0008,
      outputCostPer1K: 0.004,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    // OpenAI models via OpenRouter
    {
      id: 'openai/gpt-4o',
      name: 'GPT-4o',
      provider: 'openrouter',
      contextWindow: 128000,
      maxOutputTokens: 16384,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'code'],
      inputCostPer1K: 0.005,
      outputCostPer1K: 0.015,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    {
      id: 'openai/gpt-4o-mini',
      name: 'GPT-4o Mini',
      provider: 'openrouter',
      contextWindow: 128000,
      maxOutputTokens: 16384,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'code'],
      inputCostPer1K: 0.00015,
      outputCostPer1K: 0.0006,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    // Google models via OpenRouter
    {
      id: 'google/gemini-2.0-flash',
      name: 'Gemini 2.0 Flash',
      provider: 'openrouter',
      contextWindow: 1000000,
      maxOutputTokens: 8192,
      capabilities: ['chat', 'streaming', 'vision', 'function_calling', 'code'],
      inputCostPer1K: 0.0001,
      outputCostPer1K: 0.0004,
      supportsStreaming: true,
      supportsVision: true,
      supportsFunctionCalling: true,
    },
    // DeepSeek models
    {
      id: 'deepseek/deepseek-r1',
      name: 'DeepSeek R1',
      provider: 'openrouter',
      contextWindow: 64000,
      maxOutputTokens: 8192,
      capabilities: ['chat', 'streaming', 'reasoning', 'code'],
      inputCostPer1K: 0.00055,
      outputCostPer1K: 0.00219,
      supportsStreaming: true,
      supportsVision: false,
      supportsFunctionCalling: false,
    },
  ];

  async *chat(
    messages: ChatMessage[],
    model: string,
    options: ChatOptions
  ): AsyncGenerator<StreamChunk, void, unknown> {
    if (!this.isConfigured) {
      yield {
        type: 'error',
        error: new ProviderError(
          'OpenRouter API key not configured',
          'NOT_CONFIGURED',
          this.id
        ),
      };
      return;
    }

    // Convert messages to OpenAI format (OpenRouter uses OpenAI-compatible API)
    const formattedMessages = messages.map(m => ({
      role: m.role,
      content: m.content,
    }));

    // Add system prompt as first message if provided
    if (options.systemPrompt) {
      formattedMessages.unshift({
        role: 'system',
        content: options.systemPrompt,
      });
    }

    const body: Record<string, unknown> = {
      model,
      messages: formattedMessages,
      stream: options.stream !== false,
    };

    // Add optional parameters
    if (options.maxTokens !== undefined) {
      body.max_tokens = options.maxTokens;
    }
    if (options.temperature !== undefined) {
      body.temperature = options.temperature;
    }
    if (options.topP !== undefined) {
      body.top_p = options.topP;
    }

    try {
      const response = await fetch(`${API_URLS.OPENROUTER}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
          'HTTP-Referer': 'https://thea.app',
          'X-Title': `${APP_INFO.NAME} ${APP_INFO.VERSION} (${APP_INFO.PLATFORM})`,
        },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(STREAM_CONFIG.TIMEOUT),
      });

      if (!response.ok) {
        const errorText = await response.text();
        let errorMessage = `API error: ${response.status}`;

        try {
          const errorJson = JSON.parse(errorText);
          errorMessage = errorJson.error?.message || errorMessage;
        } catch {
          // Use default message
        }

        yield {
          type: 'error',
          error: new ProviderError(
            errorMessage,
            this.mapStatusToErrorCode(response.status),
            this.id,
            response.status
          ),
        };
        return;
      }

      // Handle streaming response
      if (options.stream !== false) {
        yield* this.parseSSEStream(response, this.parseOpenAIEvent.bind(this));
      } else {
        // Non-streaming response
        const data = await response.json();
        const content = data.choices?.[0]?.message?.content || '';

        yield { type: 'content', content };
        yield {
          type: 'done',
          finishReason: data.choices?.[0]?.finish_reason,
          usage: data.usage
            ? {
                promptTokens: data.usage.prompt_tokens || 0,
                completionTokens: data.usage.completion_tokens || 0,
                totalTokens: data.usage.total_tokens || 0,
              }
            : undefined,
        };
      }
    } catch (error) {
      if (error instanceof ProviderError) {
        yield { type: 'error', error };
      } else if (error instanceof Error) {
        yield {
          type: 'error',
          error: new ProviderError(error.message, 'NETWORK_ERROR', this.id),
        };
      }
    }
  }

  /**
   * Parse OpenAI-format SSE event
   */
  private parseOpenAIEvent(data: string): StreamChunk | null {
    const event = JSON.parse(data);

    // Check for content delta
    const delta = event.choices?.[0]?.delta;
    if (delta?.content) {
      return { type: 'content', content: delta.content };
    }

    // Check for finish
    const finishReason = event.choices?.[0]?.finish_reason;
    if (finishReason) {
      return {
        type: 'done',
        finishReason,
        usage: event.usage
          ? {
              promptTokens: event.usage.prompt_tokens || 0,
              completionTokens: event.usage.completion_tokens || 0,
              totalTokens: event.usage.total_tokens || 0,
            }
          : undefined,
      };
    }

    return null;
  }

  /**
   * Map HTTP status to error code
   */
  private mapStatusToErrorCode(status: number): ProviderError['code'] {
    switch (status) {
      case 401:
        return 'INVALID_API_KEY';
      case 429:
        return 'RATE_LIMITED';
      case 400:
        return 'INVALID_RESPONSE';
      case 404:
        return 'MODEL_NOT_FOUND';
      default:
        return status >= 500 ? 'SERVER_ERROR' : 'NETWORK_ERROR';
    }
  }
}

File: Shared/AI/Providers/OpenRouterProvider.swift
import Foundation
import OSLog

final class OpenRouterProvider: AIProvider, Sendable {
    private let logger = Logger(subsystem: "ai.thea.app", category: "OpenRouterProvider")

    let metadata = ProviderMetadata(
        name: "openrouter",
        displayName: "OpenRouter",
        logoURL: URL(string: "https://openrouter.ai/favicon.ico"),
        // swiftlint:disable:next force_unwrapping
        websiteURL: URL(string: "https://openrouter.ai")!,
        // swiftlint:disable:next force_unwrapping
        documentationURL: URL(string: "https://openrouter.ai/docs")!
    )

    // periphery:ignore - Reserved: capabilities property — reserved for future feature activation
    let capabilities = ProviderCapabilities(
        supportsStreaming: true,
        // periphery:ignore - Reserved: capabilities property reserved for future feature activation
        supportsVision: true,
        supportsFunctionCalling: true,
        supportsWebSearch: false,
        maxContextTokens: 200_000, // Depends on model
        maxOutputTokens: 16384,
        supportedModalities: [.text, .image]
    )

    private let apiKey: String
    private let baseURL = "https://openrouter.ai/api/v1"

    init(apiKey: String) {
        self.apiKey = apiKey
    }

    // MARK: - Validation

    // periphery:ignore - Reserved: validateAPIKey(_:) instance method — reserved for future feature activation
    func validateAPIKey(_ key: String) async throws -> ValidationResult {
        // periphery:ignore - Reserved: validateAPIKey(_:) instance method reserved for future feature activation
        guard let url = URL(string: "\(baseURL)/models") else {
            return .failure("Invalid API URL configuration")
        }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse {
                if httpResponse.statusCode == 200 {
                    return .success()
                } else if httpResponse.statusCode == 401 {
                    return .failure("Invalid API key")
                } else {
                    return .failure("Unexpected status code: \(httpResponse.statusCode)")
                }
            }
            return .failure("Invalid response")
        } catch {
            return .failure("Network error: \(error.localizedDescription)")
        }
    }

    // MARK: - Chat

    func chat(
        messages: [AIMessage],
        model: String,
        stream: Bool
    ) async throws -> AsyncThrowingStream<ChatResponse, Error> {
        try await chatWithOptions(messages: messages, model: model, stream: stream)
    }

    // periphery:ignore - Reserved: chatAdvanced(messages:model:options:) instance method reserved for future feature activation
    /// Advanced chat with Anthropic-specific features forwarded through OpenRouter.
    /// When routing to a Claude model, forwards: system prompts, cache control, thinking, effort.
    func chatAdvanced(
        messages: [AIMessage],
        model: String,
        options: AnthropicChatOptions
    ) async throws -> AsyncThrowingStream<ChatResponse, Error> {
        try await chatWithOptions(messages: messages, model: model, stream: options.stream, anthropicOptions: options)
    }

    private func chatWithOptions(
        messages: [AIMessage],
        model: String,
        stream: Bool,
        anthropicOptions: AnthropicChatOptions? = nil
    ) async throws -> AsyncThrowingStream<ChatResponse, Error> {
        let isClaudeModel = model.contains("claude") || model.contains("anthropic")

        // Build messages — extract system messages separately for Claude
        var systemMessages: [AIMessage] = []
        var chatMessages: [AIMessage] = []
        for msg in messages {
            if msg.role == .system {
                systemMessages.append(msg)
            } else {
                chatMessages.append(msg)
            }
        }

        let openRouterMessages: [[String: Any]] = chatMessages.map { msg in
            [
                "role": convertRole(msg.role),
                "content": msg.content.textValue
            ]
        }

        var requestBody: [String: Any] = [
            "model": model,
            "messages": openRouterMessages,
            "stream": stream
        ]

        // Forward Anthropic-specific features for Claude models
        if isClaudeModel {
            // System prompt: combine system messages + options system prompt
            var systemParts: [String] = systemMessages.map(\.content.textValue)
            if let optionSystem = anthropicOptions?.systemPrompt {
                systemParts.append(optionSystem)
            }

            if !systemParts.isEmpty {
                let systemText = systemParts.joined(separator: "\n\n")

                // Apply cache control if specified
                if let cacheControl = anthropicOptions?.cacheControl {
                    requestBody["system"] = [[
                        "type": "text",
                        "text": systemText,
                        "cache_control": [
                            "type": "ephemeral",
                            "ttl": cacheControl.ttl
                        ]
                    ]]
                } else {
                    requestBody["system"] = systemText
                }
            }

            // Thinking configuration
            if let thinking = anthropicOptions?.thinking, thinking.enabled {
                requestBody["thinking"] = [
                    "type": "enabled",
                    "budget_tokens": thinking.budgetTokens
                ]
            }

            // Effort level (Opus 4.5)
            if let effort = anthropicOptions?.effort {
                requestBody["output_config"] = [
                    "effort": effort.rawValue
                ]
            }

            // Max tokens
            if let maxTokens = anthropicOptions?.maxTokens {
                requestBody["max_tokens"] = maxTokens
            }

            // Context management (auto-clear old tool results)
            if let contextMgmt = anthropicOptions?.contextManagement {
                var edits: [[String: Any]] = []
                for edit in contextMgmt.edits {
                    var editDict: [String: Any] = [
                        "type": edit.type.rawValue,
                        "trigger": ["input_tokens": edit.trigger.inputTokens]
                    ]
                    if let keep = edit.keep {
                        editDict["keep"] = keep
                    }
                    if let exclude = edit.excludeTools {
                        editDict["exclude_tools"] = exclude
                    }
                    edits.append(editDict)
                }
                requestBody["context_management"] = ["edits": edits]
            }

            // Server tools (web search, web fetch)
            if let serverTools = anthropicOptions?.serverTools {
                var tools: [[String: Any]] = []
                for tool in serverTools {
                    tools.append(tool.toolDefinition)
                }
                requestBody["tools"] = tools
            }

            // Provider-specific routing preferences
            requestBody["provider"] = [
                "order": ["Anthropic"],
                "allow_fallbacks": false
            ]
        } else {
            // Non-Claude: include system messages inline
            if !systemMessages.isEmpty {
                let allMessages = systemMessages + chatMessages
                requestBody["messages"] = allMessages.map { msg in
                    [
                        "role": convertRole(msg.role),
                        "content": msg.content.textValue
                    ] as [String: Any]
                }
            }
        }

        guard let url = URL(string: "\(baseURL)/chat/completions") else {
            throw OpenRouterError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("https://theathe.app", forHTTPHeaderField: "HTTP-Referer")
        request.setValue("Thea", forHTTPHeaderField: "X-Title")
        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)

        if stream {
            let requestCopy = request
            return AsyncThrowingStream { continuation in
                Task { @Sendable in
                    do {
                        let (asyncBytes, response) = try await URLSession.shared.bytes(for: requestCopy)

                        guard let httpResponse = response as? HTTPURLResponse,
                              httpResponse.statusCode == 200
                        else {
                            throw OpenRouterError.invalidResponse
                        }

                        var accumulatedText = ""

                        for try await line in asyncBytes.lines {
                            if line.hasPrefix("data: ") {
                                let jsonString = String(line.dropFirst(6))

                                if jsonString == "[DONE]" {
                                    let finalMessage = AIMessage(
                                        id: UUID(),
                                        conversationID: messages.first?.conversationID ?? UUID(),
                                        role: .assistant,
                                        content: .text(accumulatedText),
                                        timestamp: Date(),
                                        model: model
                                    )
                                    continuation.yield(.complete(finalMessage))
                                    break
                                }

                                guard let data = jsonString.data(using: .utf8) else { continue }
                                let json: [String: Any]
                                do {
                                    guard let parsed = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }
                                    json = parsed
                                } catch {
                                    logger.debug("Skipped malformed SSE line: \(error.localizedDescription)")
                                    continue
                                }
                                guard let choices = json["choices"] as? [[String: Any]],
                                      let delta = choices.first?["delta"] as? [String: Any],
                                      let content = delta["content"] as? String
                                else {
                                    continue
                                }

                                accumulatedText += content
                                continuation.yield(.delta(content))
                            }
                        }

                        continuation.finish()
                    } catch {
                        continuation.yield(.error(error))
                        continuation.finish(throwing: error)
                    }
                }
            }
        } else {
            // Non-streaming
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200
            else {
                throw OpenRouterError.invalidResponse
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let choices = json["choices"] as? [[String: Any]],
                  let message = choices.first?["message"] as? [String: Any],
                  let content = message["content"] as? String
            else {
                throw OpenRouterError.noResponse
            }

            let finalMessage = AIMessage(
                id: UUID(),
                conversationID: messages.first?.conversationID ?? UUID(),
                role: .assistant,
                content: .text(content),
                timestamp: Date(),
                model: model
            )

            return AsyncThrowingStream { continuation in
                continuation.yield(.complete(finalMessage))
                continuation.finish()
            }
        }
    }

    // MARK: - Models

    func listModels() async throws -> [ProviderAIModel] {
        guard let url = URL(string: "\(baseURL)/models") else {
            throw OpenRouterError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        let (data, _) = try await URLSession.shared.data(for: request)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let modelsData = json["data"] as? [[String: Any]]
        else {
            throw OpenRouterError.noModels
        }

        return modelsData.compactMap { modelData in
            guard let id = modelData["id"] as? String,
                  let name = modelData["name"] as? String
            else {
                return nil
            }

            let contextWindow = modelData["context_length"] as? Int ?? 128_000
            let pricing = modelData["pricing"] as? [String: Any]
            let inputPrice = (pricing?["prompt"] as? String).flatMap { Double($0) }.map { Decimal($0) * 1_000_000 } ?? 0
            let outputPrice = (pricing?["completion"] as? String).flatMap { Double($0) }.map { Decimal($0) * 1_000_000 } ?? 0

            return ProviderAIModel(
                id: id,
                name: name,
                description: modelData["description"] as? String,
                contextWindow: contextWindow,
                maxOutputTokens: 16384,
                inputPricePerMillion: inputPrice,
                outputPricePerMillion: outputPrice,
                supportsVision: id.contains("vision") || id.contains("gpt-4o") || id.contains("claude"),
                supportsFunctionCalling: true
            )
        }
    }

    // MARK: - Helpers

    private func convertRole(_ role: MessageRole) -> String {
        switch role {
        case .user: "user"
        case .assistant: "assistant"
        case .system: "system"
        }
    }
}

// MARK: - Errors

enum OpenRouterError: Error, LocalizedError {
    case invalidResponse
    case noResponse
    case noModels

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            "Invalid response from OpenRouter"
        case .noResponse:
            "No response from OpenRouter"
        case .noModels:
            "No models available from OpenRouter"
        }
    }
}
